File: sdks/node/examples/direct.ts
// patala DIRECT mode on Node — libpatala_ffi inside this process, over the C
// ABI in patala-ffi/include/patala.h.
//
//   cargo build -p patala-ffi --release        # from the repo root
//   npm install && npm install koffi
//   npm run example:direct
//
// EVERY LINE OF THIS RUNS ON MockRail. patala is a payments library, and an
// example that moves real value is not an example: the mock rail is
// deterministic, needs no credentials, opens no socket, and is present in every
// build — so a full charge -> verify round trip is reachable before a single
// secret exists.
//
// Nothing here needs the network, and nothing here is allowed to want it.

import { execFileSync } from "node:child_process";
import * as fs from "node:fs";

import { abiCheck, abiVersion, Rail, resolveLibrary, type Receipt } from "../direct.js";

/**
 * How many threads this process has right now.
 *
 * Present because patala's headline claim over the Go-based C ABIs in this
 * suite — llmux, openrate — is that loading it starts nothing. A claim like
 * that should be measured in the example, not asserted in a README.
 */
function threads(): number {
  if (process.platform === "linux") return fs.readdirSync("/proc/self/task").length;
  if (process.platform === "darwin") {
    return execFileSync("ps", ["-M", String(process.pid)], { encoding: "utf8" }).trim().split("\n").length - 1;
  }
  return -1;
}

const before = threads();

console.log(`node        ${process.version} on ${process.platform}/${process.arch}`);
console.log(`library     ${resolveLibrary()}`);
console.log(`bytes       ${String(fs.statSync(resolveLibrary()).size)}`);
console.log(`abi         ${abiVersion()}`);
console.log(`threads     ${String(before)} before dlopen -> ${String(threads())} after\n`);

// patala_abi_check, not a comparison written here: the ABI exports the check so
// that twelve bindings do not each reimplement — and each forget — it.
abiCheck(abiVersion());
try {
  abiCheck("0.0.0-not-this-one");
  console.log("version     UNEXPECTED: a wrong version passed the check");
} catch (e) {
  console.log(`version     ${e instanceof Error ? e.message : "non-Error thrown"}\n`);
}

// ===========================================================================
// 1. A rail, and what it says about itself
// ===========================================================================
{
  // `using` closes the handle on every exit path out of this block, throw
  // included. Without it, an exception between open and close leaks a rail.
  using rail = Rail.open({ rail: "mock", currencies: ["USDC"], fee_minor: 25 });
  console.log(`rail        handle ${String(rail.handle)}, id ${rail.id().rail_id}`);

  const caps = rail.capabilities();
  // Decide your whole UX from `class`, without knowing which provider answered.
  // NonCustodialFinal means a wallet address and a signed, irreversible
  // receipt; CustodialReversible would mean a card form and a refundable
  // pending state. It is not a bool because those are not two shades of one
  // thing.
  console.log(`caps        ${caps.class}, reversible ${String(caps.reversible)}, holds_funds ${String(caps.holds_funds)}`);
  console.log(`            currencies ${caps.currencies.join(", ")}, settlement ${JSON.stringify(caps.settlement)}\n`);

  // =========================================================================
  // 2. quote, charge, verify — and the receipt IS the entitlement
  // =========================================================================
  const req = {
    amount_minor: 1250,
    currency: "USDC",
    destination: "mock:wallet:alice",
    reference: "order-1",
  };

  // Integer minor units on both sides. 1250 USDC base units, plus a 25 fee.
  // Never a float, anywhere.
  const q = rail.quote(req);
  console.log(`quote       ${String(q.amount_minor)} + ${String(q.fee_minor)} fee = ${String(q.total_minor)} ${q.currency}`);
  console.log(`            expires_at_unix ${String(q.expires_at_unix)}`);

  const receipt = rail.charge(req);
  console.log(`charge      ${String(receipt.amount_minor)} ${receipt.currency} for ${receipt.reference}`);
  console.log(`            proof ${String(receipt.proof.length)} bytes, settled_at_unix ${String(receipt.settled_at_unix)}`);

  // THIS is the entitlement check — not `charge` having returned. Store the
  // receipt, hand it back later, and gate on this.
  console.log(`verify      ${JSON.stringify(rail.verify(receipt))}`);

  // A tampered receipt verifies FALSE, and false is a RESULT, not an error.
  // The distinction is the whole point: if this arrived as a thrown error, a
  // caller with a retry loop would treat an unpaid order as a transient
  // failure and eventually as an entitlement.
  const tampered: Receipt = { ...receipt, amount_minor: receipt.amount_minor * 100 };
  const verdict = rail.verify(tampered);
  console.log(`tampered    ${JSON.stringify(verdict)} — a result, not a thrown error`);
  console.log(`            typeof ${typeof verdict.valid}, so \`if (v.valid)\` is the only correct gate\n`);

  // =========================================================================
  // 3. validate-destination — the offline pre-flight, which never fails
  // =========================================================================
  // Five verdicts, all reachable on the mock rail, all returned as answers.
  // "I cannot check this" is `Unknown`, not an exception, because an error is
  // too easy to swallow on the one question that decides where money goes.
  for (const dest of ["mock:wallet:alice", "mock:program:vault", "stellar:wallet:alice", "nonsense"]) {
    const v = rail.validateDestination(dest);
    console.log(`destination ${dest.padEnd(21)} ${v.status.padEnd(18)} refusal=${String(v.is_refusal)} confirm=${String(v.human_must_confirm)}`);
  }
  // human_must_confirm is true even on StructurallyValid, and this is the
  // sentence to show that human. `caveat` returns the same text with no verdict
  // attached, for the form where the address is first asked for.
  console.log(`caveat      ${rail.caveat().exchange_deposit_caveat.slice(0, 96)}…\n`);

  // =========================================================================
  // 4. What this build cannot do, said by name
  // =========================================================================
  // The mock rail has no push delivery, so it refuses rather than inventing an
  // event that a caller might gate entitlement on.
  try {
    rail.webhook({ body: "{}", headers: {}, now_unix: 1_700_000_000 });
    console.log("webhook     UNEXPECTED: the mock rail produced an event");
  } catch (e) {
    console.log(`webhook     ${e instanceof Error ? e.message : "non-Error thrown"}`);
  }

  // A default cdylib links no fiat adapters, and says which feature is missing
  // rather than answering with an empty list.
  try {
    console.log(`providers   ${JSON.stringify(rail.providers().providers)}`);
  } catch (e) {
    console.log(`providers   ${e instanceof Error ? e.message : "non-Error thrown"}`);
  }

  // An unknown method is a clean error naming the closed set.
  try {
    rail.call("refund" as never);
  } catch (e) {
    console.log(`unknown     ${e instanceof Error ? e.message : "non-Error thrown"}\n`);
  }
}

// ===========================================================================
// 5. Your error path, on demand
// ===========================================================================
{
  // `failing: true` is a mock rail where every operation fails — so the branch
  // your code takes when a processor is down is reachable offline, in a test,
  // without waiting for a real outage.
  using broken = Rail.open({ rail: "mock", failing: true });
  try {
    broken.charge({ amount_minor: 1, currency: "USDC", destination: "mock:wallet:alice", reference: "r" });
    console.log("failing     UNEXPECTED: a failing rail charged");
  } catch (e) {
    console.log(`failing     ${e instanceof Error ? e.message : "non-Error thrown"}`);
  }

  // A rail with destination checks off answers Unknown to everything — the
  // offline stand-in for a fiat rail, whose destination is an opaque
  // processor-side token. It exists so the "a human must decide" branch of a
  // payout UI is reachable in the default build.
  using opaque = Rail.open({ rail: "mock", destination_checks: false });
  const v = opaque.validateDestination("acct_1234567890");
  // Unknown is NOT "fine". It is "a human must decide", which is why
  // human_must_confirm is set and is_refusal is not: nothing was established
  // either way.
  console.log(`opaque      ${v.status}, refusal=${String(v.is_refusal)} confirm=${String(v.human_must_confirm)} — a fiat rail's honest answer`);
}

// A closed handle is a clean error, never a crash: handles are registry
// integers and are never reused, so a use-after-close says so rather than
// landing on whatever rail took the slot.
const stale = Rail.open();
stale.close();
stale.close(); // idempotent, as patala_close is
try {
  stale.id();
} catch (e) {
  // This particular message is this binding's own guard, which fires before the
  // ABI is reached. The ABI is equally safe on its own: handles are registry
  // integers, never pointers, and are never reused, so a closed or invented one
  // is `patala: unknown handle N` rather than a segfault on someone else's rail.
  console.log(`use-after   ${e instanceof Error ? e.message : "non-Error thrown"}`);
}
console.log(`threads     ${String(threads())} after every synchronous call above\n`);

// ===========================================================================
// 6. Off the main thread — which is patala-specific
// ===========================================================================
// callAsync runs patala_call on a libuv threadpool thread.
//
// It buys nothing on the mock rail, and the honest measurement below says so:
// the call finishes in microseconds, so a 1 ms timer racing it barely ticks
// either way. The point is structural — a real rail's `charge` is a network
// round trip to Solana, Stellar or a processor, and that is the one you do not
// want on the event loop.
//
// The reason this method exists here and not in llmux's or openrate's Node
// SDKs: the identical construction against their Go libraries answers correctly
// and then hangs the process forever. patala is Rust and puts no runtime in the
// host, so there is no thread stuck inside one at exit. README.md has both
// controls, run side by side.
{
  using rail = Rail.open();
  let ticks = 0;
  const timer = setInterval(() => ticks++, 1);
  const began = performance.now();
  const r = await rail.callAsync("charge", {
    amount_minor: 500,
    currency: "USDC",
    destination: "mock:wallet:bob",
    reference: "order-2",
  });
  const took = performance.now() - began;
  clearInterval(timer);
  console.log(`async       charged ${String(r.amount_minor)} ${r.currency} off-thread in ${took.toFixed(2)} ms`);
  console.log(`            the event loop ticked ${String(ticks)}x meanwhile — low because there was no time to tick in`);
  console.log(`async       verify ${JSON.stringify(await rail.callAsync("verify", r))}`);
}

// And now the count DOES move — by exactly libuv's default threadpool size.
// Those four threads are Node's, created lazily on the first threadpool
// request; `await fs.promises.readFile(...)` in a process that has never heard
// of patala moves this number identically. patala still started none.
console.log(`threads     ${String(before)} at startup -> ${String(threads())} after the off-thread call`);
console.log("            the difference is libuv's threadpool (UV_THREADPOOL_SIZE, default 4), not patala's:");
console.log("            no GC, no scheduler, no signal handler, no thread of its own. That is the Rust dividend.");

File: sdks/rust/examples/direct.rs
//! patala from Rust, **direct** — in-process, no FFI, no shared library.
//!
//! There is nothing to load and nothing to locate. `patala-core` is the core;
//! a Rust host does not reach it through a binding, it *is* the host. Every
//! other language in `sdks/` is trying to get back to what this file has for
//! free: real types, real errors, real ownership.
//!
//! Run it:
//!
//! ```text
//! cd sdks/rust && cargo run --example direct
//! ```
//!
//! Everything here drives `MockRail` — deterministic, offline, no credentials.
//! This is a payments library; an example that moves real value is not an
//! example.

use std::time::Instant;

use patala_core::{
    DestinationStatus, Error, FailoverRail, MockRail, PayRequest, PaymentRail, RailClass,
    Settlement,
};

/// Money is integer minor units plus a currency string. Never a float, and
/// never anywhere in this file — including in the printing.
fn minor(amount_minor: u64, currency: &str) -> String {
    format!(
        "{}.{:02} {}",
        amount_minor / 100,
        amount_minor % 100,
        currency
    )
}

#[tokio::main]
async fn main() {
    println!("patala direct — in-process, no FFI, no shared library");

    // ---------------------------------------------------------------- rail
    // No handle, no `open`, no version probe against a library that may be
    // stale on the load path: the rail is a value, and the compiler already
    // checked that this crate and patala-core agree about its shape.
    let rail = MockRail::new(
        "mock",
        RailClass::NonCustodialFinal,
        vec!["USDC".into(), "USD".into()],
    );
    println!("rail:      {}", rail.id());

    // ------------------------------------------------------- capabilities
    // The settlement class is a `RailClass`, not a bool and not a string, so
    // this `match` is exhaustive and a new class would be a compile error
    // here rather than a wrong UX in production. That is the thing direct
    // mode buys that no JSON boundary can give back.
    let caps = rail.capabilities();
    let ux = match caps.class {
        RailClass::CustodialReversible => "card form, refundable pending state",
        RailClass::NonCustodialFinal => "wallet address, signed final receipt",
    };
    println!(
        "caps:      {:?} / {ux}\n           settlement={:?} holds_funds={} reversible={} currencies={:?}",
        caps.class, caps.settlement, caps.holds_funds, caps.reversible, caps.currencies,
    );
    assert!(!caps.holds_funds, "patala itself never holds funds");
    assert_eq!(caps.settlement, Settlement::Instant);

    // -------------------------------------------------------------- quote
    let req = PayRequest {
        amount_minor: 1250, // 12.50 USDC
        currency: "USDC".into(),
        destination: "mock:wallet:alice".into(),
        reference: "order-1".into(),
    };

    let quote = rail.quote(&req).await.expect("quote");
    println!(
        "quote:     {} + {} fee = {} (expires {}s)",
        minor(quote.amount_minor, &quote.currency),
        minor(quote.fee_minor, &quote.currency),
        minor(quote.total_minor, &quote.currency),
        quote.expires_at_unix.saturating_sub(now_unix()),
    );

    // ------------------------------------------------- charge -> verify
    let t0 = Instant::now();
    let receipt = rail.charge(&req).await.expect("charge");
    let charged_in = t0.elapsed();
    println!(
        "charge:    {} ref={} rail={} proof={}B  [{charged_in:?}]",
        minor(receipt.amount_minor, &receipt.currency),
        receipt.reference,
        receipt.rail_id,
        receipt.proof.len(),
    );

    // `charge` having returned `Ok` is NOT the entitlement. The receipt is,
    // and `verify` is what re-derives that it still holds.
    let valid = rail.verify(&receipt).await.expect("verify");
    println!("verify:    Ok({valid})  <- gate entitlement on exactly this");
    assert!(valid);

    // A tampered receipt is `Ok(false)`, not `Err`. The distinction is the
    // whole fail-closed contract: `Err` is "I could not check", which you may
    // retry; `Ok(false)` is "I checked, and it does not hold", which you must
    // not. In Rust these cannot be confused — they are different variants.
    let mut tampered = receipt.clone();
    tampered.amount_minor = 999_999;
    match rail.verify(&tampered).await {
        Ok(false) => println!("tampered:  Ok(false) — a refusal is DATA, not an error"),
        Ok(true) => panic!("a tampered receipt verified true"),
        Err(e) => panic!("tampering must not become an error: {e}"),
    }

    // ------------------------------------------------- destination check
    // Offline, pure, and it never returns a Result — "I cannot check this" is
    // a verdict, because a caller must handle it as carefully as a refusal.
    for dest in [
        "mock:wallet:alice",
        "stellar:wallet:alice",
        "not-an-address",
    ] {
        let v = rail.validate_destination(dest);
        println!(
            "dest:      {dest:<22} {:?}{}",
            v.status,
            if v.is_refusal() {
                "  (DO NOT SEND)"
            } else {
                ""
            },
        );
        // True on EVERY verdict, including StructurallyValid: patala does not
        // detect exchange-owned addresses and will not guess.
        assert!(v.human_must_confirm);
        assert!(!v.exchange_deposit_caveat.is_empty());
    }
    let ok = rail.validate_destination("mock:wallet:alice");
    assert_eq!(ok.status, DestinationStatus::StructurallyValid);

    // A rail whose destination is an opaque processor-side token — the fiat
    // shape — answers Unknown rather than erroring, and Unknown is not a
    // green light.
    let opaque = MockRail::new("mock", RailClass::NonCustodialFinal, vec!["USD".into()])
        .without_destination_checks();
    let u = opaque.validate_destination("cus_opaque_token");
    println!(
        "dest:      {:<22} {:?}  is_refusal={} — and not an approval either",
        "cus_opaque_token",
        u.status,
        u.is_refusal()
    );
    assert_eq!(u.status, DestinationStatus::Unknown);

    // ---------------------------------------------------- the error path
    // Typed errors, matched by variant, with the offending value in the
    // message. Nothing here is a string to grep.
    let eur = PayRequest {
        currency: "EUR".into(),
        ..req.clone()
    };
    match rail.charge(&eur).await {
        Err(Error::InvalidRequest(msg)) => println!("refused:   Error::InvalidRequest({msg:?})"),
        other => panic!("expected a refusal naming the currency, got {other:?}"),
    }

    // Paying a customer back on a final rail is not a reversal — it is a
    // second charge to an address the CUSTOMER supplies. The rail says so in
    // the type system rather than pretending.
    match rail.refund(&receipt).await {
        Err(Error::Unsupported(op)) => {
            println!("refund:    Error::Unsupported({op:?}) — see docs/compensating-payments.md")
        }
        other => panic!("a NonCustodialFinal rail must not reverse: {other:?}"),
    }

    // The mock has no processor, so it invents no webhook event.
    let delivery = patala_core::WebhookDelivery::new(b"{}".to_vec(), 1_700_000_000);
    match rail.verify_webhook(&delivery).await {
        Err(Error::Unsupported(op)) => println!("webhook:   Error::Unsupported({op:?})"),
        other => panic!("the mock must not invent an event: {other:?}"),
    }

    // ------------------------------------------------ what only Rust gets
    // `FailoverRail` composes `Box<dyn PaymentRail>`s and refuses to cross the
    // settlement-class boundary without an explicit opt-in. It is reachable
    // from Rust and from nowhere else: the C ABI and the sidecar both hand
    // out one rail at a time, so a caller in C or over HTTP has to rebuild
    // this — including the guard — by hand.
    let chain = FailoverRail::new(vec![
        Box::new(
            MockRail::new("primary", RailClass::NonCustodialFinal, vec!["USDC".into()]).failing(),
        ),
        Box::new(MockRail::new(
            "backup",
            RailClass::NonCustodialFinal,
            vec!["USDC".into()],
        )),
    ]);
    let r = chain.charge(&req).await.expect("failover");
    println!("failover:  primary failed -> settled on {:?}", r.rail_id);

    let crossing = FailoverRail::new(vec![
        Box::new(
            MockRail::new("primary", RailClass::NonCustodialFinal, vec!["USDC".into()]).failing(),
        ),
        Box::new(MockRail::new(
            "card",
            RailClass::CustodialReversible,
            vec!["USDC".into()],
        )),
    ]);
    match crossing.charge(&req).await {
        Err(Error::CrossClassFailover { from, to }) => {
            println!("guard:     refused {from:?} -> {to:?} — the payer was promised one of these")
        }
        other => panic!("the class boundary must not be crossed silently: {other:?}"),
    }

    println!("\nOK — offline, MockRail only, no value moved.");
}

fn now_unix() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}
